### ISSUE TITLE
New rule proposal: no-unassigned-vars

### ISSUE THREAD

**Issue by mdjermanovic** [eslint/eslint#12497]
**Please describe what the rule should do:**

Reports `var` and `let` variables that are never assigned a value. In other words, variables that have zero write references. 

**What category of rule is this? (place an "X" next to just one item)**

[X] Warns about a potential error (problem)

**Provide 2-3 code examples that this rule will warn about:**

```js
var obj;
obj.x = 1;
obj.y = 2;

let a, b, c;
if (something) {
    a = 0;
    b = false;
} else {
    a = 1;
    b = true;
}
f(a, b, c);
```

**Why should this rule be included in ESLint (instead of a plugin)?**

This rule looks like a nice addition to the `Variables` category of rules, a complement to `no-unused-vars` to detect possible errors in the code.

An unassigned variable is a possible error because it always has `undefined` value. Even if that is the intention, it's more clear to simply inline `undefined` instead of reading from the variable, or to make the intention explicit with `const foo = undefined` instead of just `let foo` which looks more like an error in the code where it's never assigned a value.

`no-undef-init` does not disallow `const`.

`prefer-const` does not report these cases.

To avoid overlapping with `no-unused-vars`, this rule would report only variables that have at least 1 read reference.

**Are you willing to submit a pull request to implement this rule?**

Yes

**Comment by mysticatea** [comment]
It would be great if the rule checks each code path.

```js
let a, b, c;
if (something) {
    a = 0;
    b = false;
    c = ""
} else {
    a = 1;
    b = true;
}
f(a, b, c); // `c` is not initialized on a code path.
```

**Comment by mdjermanovic** [comment]
I was thinking about that, but a problem could be that leaving the `undefined` value in some code paths might be a valid use?

**Comment by mysticatea** [comment]
In that case, I think that explicit assignment (`c = undefined`) is a better choice for maintainability.

**Comment by platinumazure** [comment]
There's a part of me that wonders if user should have the option for the rule to check all code paths, vs allow the variable if it is assigned in one code path. But I don't feel strongly about this.

**Comment by mdjermanovic** [comment]
How should the rule treat variables that have a reference from a function, for instance:

```js
function f() {
    x = 1;
}

function g() {
    f();
}

someImportedFunction(g);

var x;

console.log(x); // no warning because it could be assigned?

```

This looks complicated to track, it might be better to just ignore these variables?

**Comment by mdjermanovic** [comment]
> There's a part of me that wonders if user should have the option for the rule to check all code paths, vs allow the variable if it is assigned in one code path. But I don't feel strongly about this.

I also think there should be some options. A total of zero writing references (or zero writes in all code paths that lead to a reading reference) is a very possible error so this check should be probably always enabled. Enforcing assignments in all paths is a best practice but it could be reporting a valid code in an existing codebase.

**Comment by mysticatea** [comment]
An option that allows unassigned code paths sounds good.

> How should the rule treat variables that have a reference from a function, for instance:

I think that the `x` should be warned if the code path check is enabled. It's valid but cryptic.

On the other hand, I don't think that we need the code path check for the variables that come from the out of the current (function/global) scope.

**Comment by mdjermanovic** [comment]
I'm thinking about the following configuration:

A string option which sets the 'mode' of this rule:

* `"any"` - Reports `read` references only if the variable does not have any `write` references. If the variable has at least 1 `write` reference anywhere in the code, all `read` references are valid. This mode does not check code paths at all.
* `"codePath"` - Reports a `read` reference if there is no `write` reference in at least 1 code path that leads to that `read` reference.
* `"allCodePaths"` - All code paths that lead to the `read` reference must have a `write` reference.

Code path checks do not take into account calls to inner functions (which could have `write` references).

Additional object option with one boolean option:

* `"ignoreAssignedFromInnerFunction"` - when `true`, a variable that has at least 1 `write` reference from an inner functions is ignored (all `read` references to that variable anywhere in the code are valid). This option makes sense only when the string option is set to `"codePath"` or `"allCodePaths"`.

Regardless of the configuration, `read` references in inner functions are reported only if the variable does not have any `write` references in the code.

**Comment by nzakas** [comment]
@mdjermanovic do you still intend to work on this? Technically it wouldn't qualify under our [new rules policy](https://eslint.org/blog/2020/05/changes-to-rules-policies) but since you already started on it, I think we can make an exception if you intend to follow up.

**Comment by mdjermanovic** [comment]
I'm willing to work on this, for sure! On the other hand, I believe that this still needs a lot of additional analysis & design, and would end up being a fairly complex rule, so I definitely wouldn't mind closing this in regard to the reasons for the new rules policy.

**Comment by nzakas** [comment]
Okay, let's close this then. Thanks!
